JAMES-2997 [ADR] Separate attachment content and metadata (#3289)
linagora/james-project: src/adr/0030-separate-attachment-content-and-metadata.md

# 30. Separate attachment content and metadata

Date: 2020-04-13

## Status

Accepted (lazy consensus)

## Context

> Some mailbox implementations of James stores already parsed attachment for faster retrieval.

Reviewer: s/stores/store
s/attachment/attachments ?

> Here are the POJOs representing these attachments:

Reviewer 2: the list doesn't contain POJOs representing these attachments, it's more the list of POJOs related to attachements

 - **Attachment** : holds an attachmentId, the attachment content, as well as the content type

>  - **MessageAttachment** : compose an attachment with its disposition within a message (cid, inline and name)

Reviewer: s/compose/composes

>  - **MessageAttachment** : compose an attachment with it's disposition within a message (cid, inline and name)

Reviewer 3: ```suggestion
 - **MessageAttachment** : compose an attachment with its disposition within a message (cid, inline and name)
```

>  - **Message** exposes it's list of MessageAttachment when it is read with FetchType Full.

Reviewer: s/it's/its

>  - **Message** exposes it's list of MessageAttachment when it is read with FetchType Full..

Reviewer 3: ```suggestion
 - **Message** exposes it's list of MessageAttachment when it is read with FetchType Full.
```

 - **Blob** represents some downloadable content, and can be either an attachment or a message. Blob has a byte array 
 payload too.
 

> The following classes works with the aforementioned POJOs:

Reviewer 3: ```suggestion
The following classes work with the aforementioned POJOs:
```

 
 - **AttachmentMapper** and **AttachmentManager** are responsible of storing and retrieving an attachment content.
 - **BlobManager** is used by JMAP to allow blob downloads.
 - Mailbox search exposes attachment content related criteria. These criteria are used by the JMAP protocol.

> This organisation causes attachment content to be loaded every time a message is fully read (which happens for instance

Reviewer 2: s/despite the fact that it's not needed./despite the fact that it's usually not needed.

Author: I meant it:

that it's not needed, as attachments are downloadable through a 
separate JMAP endpoint, their content is not attached to the JMAP message JSON.

Yes, it's completly useless.

Reviewer 2: I understand we don't use that in JMAP. It's not clear at this point of the ADR why we only talk about JMAP. Maybe it deserves to explain why it's the only thing that we care about.

when you open a message using JMAP) despite the fact that it's not needed.

> Also, the content being loaded "at once", we allocate memory space to store the whole attachment, which is sub-optimal.

Reviewer 2: sub-optimal is not describing the problem well enough, it could be optimal if you want to avoid keeping file descriptors open.
In our case, we want to be able to keep the consumed memory low per-message because a given server should be able to handle thousands of messages at a given time.

> To be noted that JPA and maildir mailbox implementation do not support attachment storage. To retrieve attachment of a 

Reviewer 4: s/attachment/attachments ?

> To be noted that JPA and maildir mailbox implementation do not support attachment storage. To retrieve attachment of a 

Reviewer: s/implementation/implementations

> message these implementations parse the messages to extract their attachments.

Reviewer 4: s/`message`/`message, ` ?

> messages these implementations parses the message to extract their attachments.

Reviewer 3: ```suggestion
message these implementations parse the messages to extract their attachments.
```

> message, these implementations parses the message to extract their attachments.

Reviewer 5: s/parses/parse

> Cassandra mailbox prior schema version 4 stores attachment and its metadata in the same table, but from version 5 relies 

Reviewer 3: ```suggestion
Cassandra mailbox prior schema version 4 stored attachment and its metadata in the same table, but from version 5 relies 
```

on the blobStore to store the attachment content.

## Decision

Enforce cassandra schema version to be 5 from James release 3.5.0. This allows to drop attachment management prior version
5.

We will re-organize the attachment POJOs: 

 - **Attachment** should hold an attachmentId, a content type, and a size. It will no longer hold the content.

>  - **MessageAttachment** : compose an attachment with its disposition within a message (cid, inline and name)

Reviewer: s/compose/composes

>  - **MessageAttachment** : compose an attachment with it's disposition within a message (cid, inline and name)

Reviewer 3: ```suggestion
 - **MessageAttachment** : compose an attachment with its disposition within a message (cid, inline and name)
```

 - **Blob** would no longer hold the content as a byte array but rather a content retriever (`Supplier<InputStream>`)
 - **ParsedAttachment** is the direct result of attachment parsing, and composes a **MessageAttachment** and the 
 corresponding content as byte array. This class is only relied upon when saving a message in mailbox. This is used as 
 an output of `MessageParser`.

Some adjustments are needed on class working with attachment:

>  - **AttachmentMapper** and **AttachmentManager** needs to allow from an attachmentId to retrieve the attachment content

Reviewer 3: ```suggestion
 - **AttachmentMapper** and **AttachmentManager** need to allow from an attachmentId to retrieve the attachment content
```

 as an `InputStream`. This is done through a separate `AttachmentLoader` interface.

>  - **AttachmentMapper** and **AttachmentManager** needs the Attachment and it's content to persist an attachment

Reviewer 3: ```suggestion
 - **AttachmentMapper** and **AttachmentManager** need the Attachment and its content to persist an attachment
```

>  - **AttachmentMapper** and **AttachmentManager** needs the Attachment and it's content to persist an attachment

Reviewer 5: s/it's/its

 - **MessageManager** then needs to return attachment metadata as a result of Append operation.
 - **InMemoryAttachmentMapper** needs to store attachment content separately.

>  - **MessageStorer** will take care of storing a message on the behalf of `MessageManager`. This enablesto determine if 

Reviewer 3: ```suggestion
 - **MessageStorer** will take care of storing a message on the behalf of `MessageManager`. This enables to determine if 
```

 attachment should be parsed or not on an implementation aware fashion, saving attachment parsing upon writes for JPA 
 and Maildir.
 

> Maildir and JPA no longer support attachment content loading.

Reviewer 2: why?

Author: Because it's used nowhere cf JMAP only requires attachment content loading.

This could be developped with coupling on the attachment naming strategy.

Mailbox search attachment content criteria will be supported only on implementation supporting attachment storage.

## Consequences

Users running Cassandra schema version prior version 5 will have to go through James release 3.5.0 to upgrade to a 
version after version 5 before proceeding with their update.

> We noticed performance enhancement when using IMAP FETCH and JMAP GetMessages. Runnin a Gatling test suite exercising 

Reviewer: s/Runnin/Running

> We noticed performance enhancement when using IMAP FETCH and JMAP GetMessages. Runnin g a gatling test suite exercising 

Reviewer 3: ```suggestion
We noticed performance enhancement when using IMAP FETCH and JMAP GetMessages. Runnin a Gatling test suite exercising 
```

This organisation causes attachment content to be loaded every time a message is fully read (which happens for instance
when you open a message using JMAP) despite the fact that it is not needed, as attachments are downloadable through a 
separate JMAP endpoint, their content is not attached to the JMAP message JSON.

Also, the content being loaded "at once", we allocate memory space to store the whole attachment, which is sub-optimal. We
want to keep the consumed memory low per-message because a given server should be able to handle a high number of messages 
at a given time.

To be noted that JPA and maildir mailbox implementations do not support attachment storage. To retrieve attachments of a 
message, these implementations parse the messages to extract their attachments.

Cassandra mailbox prior schema version 4 stored attachment and its metadata in the same table, but from version 5 relies 
on the blobStore to store the attachment content.

## Decision

Enforce cassandra schema version to be 5 from James release 3.5.0. This allows to drop attachment management prior version
5.

We will re-organize the attachment POJOs: 

>  - **Attachment** should hold an attachmentId, a content type, and a size. It will no longer hold the content.

Reviewer 3: that's unclear where one can find the content now

 - **MessageAttachment** : composes an attachment with it's disposition within a message (cid, inline and name)
 - **Blob** would no longer hold the content as a byte array but rather a content retriever (`Supplier<InputStream>`)
 - **ParsedAttachment** is the direct result of attachment parsing, and composes a **MessageAttachment** and the 
 corresponding content as byte array. This class is only relied upon when saving a message in mailbox. This is used as 
 an output of `MessageParser`.

Some adjustments are needed on class working with attachment:

>  - **AttachmentMapper** and **AttachmentManager** needs to allow from an attachmentId to retrieve the attachment content

Author: Thanks but it is outdated

 as an `InputStream`. This is done through a separate `AttachmentLoader` interface.

>  - **AttachmentMapper** and **AttachmentManager** needs the Attachment and its content to persist an attachment

Author: Thanks but it is outdated

 attachment should be parsed or not on an implementation aware fashion, saving attachment parsing upon writes for JPA 
 and Maildir.
 
Maildir and JPA no longer support attachment content loading. Only the JMAP protocol requires attachment content loading,

> which is not supported on top of these technologies. Attachment loading could be supported in the future via coupling on 

Reviewer 2: I don't get why we would describe a solution here (that I don't understand as-is anyway)

Author: Then I will remove the explanation
